### PR TITLE
Close channel refactor

### PR DESCRIPTION
**Pull request by kerzhner**
The refactor eliminates the action pattern previously used by protocol.

This work is a progression toward https://github.com/statechannels/statechannels/issues/3124. To tackle that issue, the first step is to isolate the defunding logic in the close channel protocol. I had trouble doing so with the current implementation. Also, we are no longer using the "action pattern" in any other protocol.

As this PR is a refactor, the PR relies on existing tests to verify correctness of changes.

### CODE REVIEW

**Comment by kerzhner** [comment]
> Is this ready for review? I see that some project management tasks in the PR checklist were not done for this PR. For instance, it is not connected to the issue it pertains to.

Yep it is ready for review. I added this PR as a blocker to https://github.com/statechannels/statechannels/issues/3124. I don't think I am missing anything else, but maybe I am...

**Comment by andrewgordstewart** [comment]
Making this _block_ #3124 is confusing to me, since it is _addressing_ #3124. So it seems to me like it should be _connected_ to #3124:

Screen Shot 2021-02-02 at 3 56 42 PM

File: packages/server-wallet/src/protocols/close-channel.ts [hunk 1]
```diff
@@ -55,70 +43,100 @@ export class ChannelCloser {
...
+            attemptAnotherProtocolStep = false;
+            await this.completeObjective(objective, protocolState, tx, response);
+          } else {
+            response.queueChannelState(protocolState.app);
             attemptAnotherProtocolStep = false;
           }
-        } else {
-          response.queueChannelState(protocolState.app);
+        } catch (error) {
+          this.logger.error({error}, 'Error handling action');
```
Comment (andrewgordstewart) [inline packages/server-wallet/src/protocols/close-channel.ts:67]: It's not really handling an action anymore, right?

**Review by andrewgordstewart — APPROVED** [review]

File: packages/server-wallet/src/protocols/close-channel.ts (hunk 1 above)
Comment (kerzhner) [reply packages/server-wallet/src/protocols/close-channel.ts]: fa753ddba

**Comment by kerzhner** [comment]
> Making this _block_ #3124 is confusing to me, since it is _addressing_ #3124. So it seems to me like it should be _connected_ to #3124:
> 
> Screen Shot 2021-02-02 at 3 56 42 PM

This PR does not resolve https://github.com/statechannels/statechannels/issues/3124. That issue should probably get reworded since its purpose is for channel closing objective and channel defunding objective (which might actually become the channel challenge objective) to share the defunding logic.

Note that this PR refactors the close channel protocol to create an avenue to share the defunding logic between any objective that needs the defunding logic.

**Pull request merged**